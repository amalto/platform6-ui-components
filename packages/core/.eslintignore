dist/
loader/

src/components.d.ts