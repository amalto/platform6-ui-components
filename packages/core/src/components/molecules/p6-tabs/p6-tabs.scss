@import '@fortawesome/fontawesome-svg-core/styles';
@import '@creativebulma/bulma-tooltip/src/sass/index';

$tabs-link-active-border-top-color: $primary;
$tabs-link-active-color: $success;

$tabs-border-bottom-color: $border !default;
$tabs-border-bottom-style: solid !default;
$tabs-border-bottom-width: 1px !default;
$tabs-link-color: $text !default;
$tabs-link-hover-border-bottom-color: $text-strong !default;
$tabs-link-hover-color: $text-strong !default;
$tabs-link-active-border-top-color: $link !default;
$tabs-link-active-color: $link !default;
$tabs-link-padding: 0.5em 1em !default;
$tabs-link-max-width: 7rem;

$tabs-boxed-link-radius: $radius !default;
$tabs-boxed-link-hover-background-color: $background !default;
$tabs-boxed-link-hover-border-bottom-color: $border !default;

$tabs-boxed-link-active-background-color: $scheme-main !default;
$tabs-boxed-link-active-border-color: $border !default;
$tabs-boxed-link-active-border-bottom-color: transparent !default;

$tabs-toggle-link-border-color: $border !default;
$tabs-toggle-link-border-style: solid !default;
$tabs-toggle-link-border-width: 1px !default;
$tabs-toggle-link-hover-background-color: $background !default;
$tabs-toggle-link-hover-border-color: $border-hover !default;
$tabs-toggle-link-radius: $radius !default;
$tabs-toggle-link-active-background-color: $link !default;
$tabs-toggle-link-active-border-color: $link !default;
$tabs-toggle-link-active-color: $link-invert !default;

:host {
  display: block;

  .tabs {
    -webkit-overflow-scrolling: touch;
    @include unselectable;

    align-items: stretch;
    display: flex;
    font-size: $size-normal;
    justify-content: space-between;
    overflow: visible;
    white-space: nowrap;
    margin-bottom: 0;
    position: relative;
    z-index: 100;

    a {
      align-items: center;
      border-bottom-color: $tabs-border-bottom-color;
      border-bottom-style: $tabs-border-bottom-style;
      border-bottom-width: $tabs-border-bottom-width;
      color: $tabs-link-color;
      display: flex;
      justify-content: center;
      margin-bottom: -#{$tabs-border-bottom-width};
      padding: $tabs-link-padding;
      vertical-align: top;
      cursor: pointer;
      text-decoration: none !important;
      position: relative;
      max-width: $tabs-link-max-width;

      .title {
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
      }

      .delete.disabled-close {
        display: none;
      }

      .delete:not(.disabled-close) {
        @include delete;
      }

      &:hover {
        border-bottom-color: $tabs-link-hover-border-bottom-color;
        color: $tabs-link-hover-color;
      }
    }

    &.is-boxed a {
      border: 1px solid transparent;
      border-radius: $tabs-boxed-link-radius $tabs-boxed-link-radius 0 0;
    }
    &.is-toggle a {
      border-color: $tabs-toggle-link-border-color;
      border-style: $tabs-toggle-link-border-style;
      border-width: $tabs-toggle-link-border-width;
      margin-bottom: 0;
      position: relative;
    }

    li {
      display: block;

      a:not(.is-active) {
        background-color: $grey-lightest;
        border-left: 1px solid $white;
        border-right: 1px solid $white;
      }

      a.is-active,
      &.is-active a {
        border-top: 2px solid $tabs-link-active-border-top-color;
        border-left: 1px solid $tabs-boxed-link-hover-border-bottom-color;
        border-right: 1px solid $tabs-boxed-link-hover-border-bottom-color;
        border-bottom-color: $white;
        color: $tabs-link-active-color;
      }
    }

    &.is-boxed a:hover {
      background-color: $tabs-boxed-link-hover-background-color;
      border-bottom-color: $tabs-boxed-link-hover-border-bottom-color;
    }

    &.is-toggle a:hover {
      background-color: $tabs-toggle-link-hover-background-color;
      border-color: $tabs-toggle-link-hover-border-color;
      z-index: 2;
    }

    li a:hover:not(.is-active) {
      border-bottom-color: $tabs-link-hover-border-bottom-color;
      color: $tabs-link-hover-color;
    }

    ul {
      align-items: center;
      border-bottom-color: $tabs-border-bottom-color;
      border-bottom-style: $tabs-border-bottom-style;
      border-bottom-width: $tabs-border-bottom-width;
      display: flex;
      flex-grow: 1;
      flex-shrink: 0;
      justify-content: flex-start;
      margin: 0;
      padding: 0;

      &.is-left {
        padding-right: 0.75em;
      }

      &.is-center {
        flex: none;
        justify-content: center;
        padding-left: 0.75em;
        padding-right: 0.75em;
      }

      &.is-right {
        justify-content: flex-end;
        padding-left: 0.75em;
      }
    }

    .icon {
      &:first-child {
        margin-right: 0.5em;
      }

      &:last-child {
        margin-left: 0.5em;
      }
    }

    // Alignment
    &.is-centered {
      ul {
        justify-content: center;
      }
    }

    &.is-right {
      ul {
        justify-content: flex-end;
      }
    }

    // Styles
    &.is-boxed li {
      &.is-active {
        a {
          background-color: $tabs-boxed-link-active-background-color;
          border-color: $tabs-boxed-link-active-border-color;
          border-bottom-color: $tabs-boxed-link-active-border-bottom-color !important;
        }
      }
    }

    &.is-fullwidth {
      li {
        flex-grow: 1;
        flex-shrink: 0;
      }
    }

    &.is-toggle {
      li {
        & + li {
          margin-left: -#{$tabs-toggle-link-border-width};
        }

        &:first-child a {
          border-top-left-radius: $tabs-toggle-link-radius;
          border-bottom-left-radius: $tabs-toggle-link-radius;
        }

        &:last-child a {
          border-top-right-radius: $tabs-toggle-link-radius;
          border-bottom-right-radius: $tabs-toggle-link-radius;
        }

        &.is-active {
          a {
            background-color: $tabs-toggle-link-active-background-color;
            border-color: $tabs-toggle-link-active-border-color;
            color: $tabs-toggle-link-active-color;
            z-index: 1;
          }
        }
      }

      ul {
        border-bottom: none;
      }

      &.is-toggle-rounded {
        li {
          &:first-child a {
            border-bottom-left-radius: $radius-rounded;
            border-top-left-radius: $radius-rounded;
            padding-left: 1.25em;
          }

          &:last-child a {
            border-bottom-right-radius: $radius-rounded;
            border-top-right-radius: $radius-rounded;
            padding-right: 1.25em;
          }
        }
      }
    }

    // Sizes
    &.is-small {
      font-size: $size-small;
    }

    &.is-medium {
      font-size: $size-medium;
    }

    &.is-large {
      font-size: $size-large;
    }
  }

  .tab-content {
    padding: $gap-size * 2 * 1px;
    border-left: 1px solid $tabs-border-bottom-color;
    border-right: 1px solid $tabs-border-bottom-color;
    border-bottom: 1px solid $tabs-border-bottom-color;
  }
}
