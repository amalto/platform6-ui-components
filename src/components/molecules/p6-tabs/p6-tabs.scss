@import "bulma/sass/layout/section";
@import "bulma/sass/components/tabs";
@import "src/styles/atoms/_anchor";

:host {
  display: block;

  .tabs {
    margin-bottom: 0;

    a {
      cursor: pointer;
      text-decoration: none !important;
    }

    li {
      a.is-active {
        border-bottom-color: $tabs-link-active-border-bottom-color;
        color: $tabs-link-active-color;
      }

      a:hover:not(.is-active) {
        border-bottom-color: $tabs-link-hover-border-bottom-color;
        color: $tabs-link-hover-color;
      }
    }

    ul {
      margin: 0;
    }
  }

  .tab-content {
    padding: $gap-size * 2 * 1px;
  }
}
