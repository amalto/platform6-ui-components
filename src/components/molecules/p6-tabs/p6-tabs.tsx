import {
  Component,
  Element,
  h,
  Host,
  Method,
  Prop,
  State,
} from "@stencil/core";
import { isEmpty } from "~utils/attribute";

export interface Tab {
  id: string;
  title: string;
}

@Component({
  tag: "p6-tabs",
  styleUrl: "p6-tabs.scss",
  shadow: true,
})
export class P6Tabs {
  @Element() host!: HTMLP6TabsElement;

  /**
   * Default tab selected.
   */
  @Prop() default: string | undefined;

  /**
   * Set selected tab.
   */
  @State() tabSelected!: string | undefined;

  /**
   * Public method to set tabSelected id.
   * Use this method by keeping a ref of the component.
   *
   * @param id Id to be selected
   */
  @Method()
  async setTabSelected(id: string): Promise<void> {
    if (this.getTabs().some((tab) => tab.id === id)) {
      this.tabSelected = this.getTabId(id);
    }
  }

  /**
   * Return the formated tab id
   *
   * @param { string } id
   */
  private getTabId = (id: string): string => {
    return `${id}-tab`;
  };

  /**
   * Set selected tab
   *
   * @param {MouseEvent} event
   */
  private handleTabSelection = (event: MouseEvent): void => {
    event.preventDefault();
    const { id } = event.currentTarget as HTMLAnchorElement;
    this.tabSelected = id;
  };

  /**
   * Get tabs from slot sttributes.
   */
  private getTabs(): Tab[] {
    return Array.from(this.host.children)
      .filter(
        (c) =>
          !isEmpty(c.getAttribute("title")) && !isEmpty(c.getAttribute("id"))
      )
      .map((c) => ({
        title: c.getAttribute("title") as string,
        id: c.getAttribute("id") as string,
      }));
  }

  private getContent(): JSX.Element {
    const child = Array.from(this.host.children).filter((c) => {
      const id = c.getAttribute("id");
      return (
        !isEmpty(c.getAttribute("title")) &&
        !isEmpty(id) &&
        this.getTabId(id as string) === this.tabSelected
      );
    })[0];

    return <div class="tab-content">{child?.innerHTML}</div>;
  }

  componentWillLoad(): void {
    this.tabSelected = this.getTabId(this.default || this.getTabs()[0]?.id);
  }

  render(): JSX.Element | null {
    const { tabSelected } = this;
    const tabs = this.getTabs();

    if (tabs.length === 0) {
      return null;
    }

    return (
      <Host>
        <div class="tabs">
          <ul>
            {tabs.map((tab) => (
              <li
                class={
                  tabSelected === this.getTabId(tab.id)
                    ? "is-active"
                    : undefined
                }
              >
                <a
                  href={`#${tab.id}`}
                  id={`${this.getTabId(tab.id)}`}
                  onClick={this.handleTabSelection}
                >
                  {tab.title}
                </a>
              </li>
            ))}
          </ul>
        </div>
        {this.getContent()}
      </Host>
    );
  }
}
