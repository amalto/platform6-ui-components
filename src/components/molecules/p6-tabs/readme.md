# p6-tabs

<!-- Auto Generated Below -->

## Properties

| Property   | Attribute  | Description           | Type                  | Default     |
| ---------- | ---------- | --------------------- | --------------------- | ----------- |
| `selected` | `selected` | Default tab selected. | `string \| undefined` | `undefined` |

## Events

| Event        | Description     | Type                              |
| ------------ | --------------- | --------------------------------- |
| `p6CloseTab` | Close tab event | `CustomEvent<{ tabId: string; }>` |

---

_Built with [StencilJS](https://stenciljs.com/)_
