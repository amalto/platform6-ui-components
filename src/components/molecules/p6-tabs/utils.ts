export interface Tab extends Element {
  id: string;
  title: string;
}

export function getTabId(id: string): string {
  return `${id}-tab`;
}

/**
 * A tab is valid only with `id` and `title` attributes.
 * @param { Element[] } tabs
 */
export function getValidTabs(tabs: Element[]): Element[] {
  return tabs.filter((tab) => "title" in tab && "id" in tab);
}
