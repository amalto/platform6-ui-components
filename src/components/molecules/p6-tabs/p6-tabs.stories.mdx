import { Story, Preview, Meta } from "@storybook/addon-docs/blocks";
import { action } from "@storybook/addon-actions";
import readme from "./readme.md";

<Meta title="Molecules/Tabs" component="p6-tabs" parameters={{ notes: readme }} />

# Tabs

`p6-tabs` provides a tabs component which display content based on the tab selected.

<Preview>
  <Story name="Default" height="300px">
    {`
      <p6-tabs default="first">
        <div id="first" title="What is Lorem Ipsum?">
          Lorem Ipsum is simply dummy text of the printing and typesetting industry.
        </div>
        <div id="second" title="What is Lorem Ipsum?">
          It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout.
        </div>
        <div id="third" title="Why do we use it?">
          It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout.
        </div>
        <div id="fourth" title="What is Lorem Ipsum?">
          There are many variations of passages of Lorem Ipsum available, but the majority have suffered alteration in some form, by injected humour, or randomised words which don't look even slightly believable.
        </div>
      </p6-tabs>
    `}
  </Story>
</Preview>
