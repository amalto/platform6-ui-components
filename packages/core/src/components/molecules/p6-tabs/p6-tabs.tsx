import { Component, ComponentInterface, Element, Event, EventEmitter, forceUpdate, h, Host, JSX, Method, State } from '@stencil/core';
import { toArray } from '../../../utils/dom';
import { isTabValid } from './utils';

@Component({
  tag: 'p6-tabs',
  styleUrl: 'p6-tabs.scss',
  shadow: true,
})
export class P6Tabs implements ComponentInterface {
  @Element() host!: HTMLP6TabsElement;

  /**
   * Close tab event
   */
  @Event({ eventName: 'p6CloseTab' }) closeTabEmitter?: EventEmitter<{ tabId: string }>;

  @State() selectedTabId: string | undefined;

  /**
   * refresh the component
   */
  @Method()
  async refresh(): Promise<void> {
    forceUpdate(this.host);
  }

  /**
   * close a tab
   * @param tabId the id of the tab to be close
   */
  @Method()
  async close(tabId: string): Promise<boolean> {
    const tabToClose = this.host.querySelector<HTMLP6TabElement>(`#${tabId}`);

    if (tabToClose !== null) {
      this.closeTab(tabToClose);
      return Promise.resolve(true);
    }
    return Promise.resolve(false);
  }

  private getTabs(): HTMLP6TabElement[] {
    return toArray(this.host.children).filter(isTabValid);
  }

  componentWillLoad(): void {
    const tabs = this.getTabs();
    const selectedTab = tabs.find(tab => tab.active);

    if (tabs.length === 0) {
      return;
    }

    if (selectedTab === undefined) {
      tabs[0].active = true;
      this.selectedTabId = tabs[0].id;
    } else {
      this.selectedTabId = selectedTab.id;
    }
  }

  componentWillRender(): void {
    this.getTabs().forEach(tab => {
      if (tab.closed) {
        this.host.removeChild(tab);
      }
    });
  }

  render(): JSX.Element | null {
    const tabs = this.getTabs();

    if (tabs.length === 0) {
      return null;
    }

    return (
      <Host>
        <div class="tabs">
          <ul>{tabs.map(tab => this.renderTab(tab))}</ul>
        </div>
        <div class="tab-content">
          <slot />
        </div>
      </Host>
    );
  }

  private renderTab(tab: HTMLP6TabElement): JSX.Element {
    return (
      <li class="has-tooltip-arrow has-tooltip-bottom" data-tooltip={tab.title}>
        <a class={tab.active ? 'is-active' : undefined} href={`#${tab.id}`} onClick={this.selectTabHandler(tab)}>
          <span class="title">{tab.title}</span>
          <span aria-hidden="true" class={{ 'delete': true, 'disabled-close': !tab.closeable }} onClick={this.closeTabHandler(tab)} />
        </a>
      </li>
    );
  }

  private selectTabHandler(selectedTab: HTMLP6TabElement): (event: MouseEvent) => void {
    return (event: MouseEvent): void => {
      event.preventDefault();

      this.getTabs().forEach(tab => {
        // eslint-disable-next-line no-param-reassign
        tab.active = tab.id === selectedTab.id;
      });
      this.selectedTabId = selectedTab.id;
    };
  }

  private closeTabHandler(closedTab: HTMLP6TabElement): (event: MouseEvent) => void {
    return (event: MouseEvent): void => {
      event.preventDefault();
      event.stopPropagation();

      this.closeTab(closedTab);
    };
  }

  private closeTab(tabToclose: HTMLP6TabElement): void {
    tabToclose.close();

    const tabs = this.getTabs();
    const closedTabIndex = tabs.findIndex(tab => tab.id === tabToclose.id);

    if (tabToclose.active && tabs[closedTabIndex - 1] !== undefined) {
      tabs[closedTabIndex - 1].active = true;
      this.selectedTabId = tabs[closedTabIndex - 1].id;
    } else {
      forceUpdate(this.host);
    }

    this.closeTabEmitter?.emit({ tabId: tabToclose.id });
  }
}
