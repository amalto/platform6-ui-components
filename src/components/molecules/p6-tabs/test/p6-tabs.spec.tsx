import { newSpecPage } from '@stencil/core/testing';
import { P6Tabs } from '../p6-tabs';

describe('p6-tabs', () => {
  it('Should be empty', async () => {
    const page = await newSpecPage({
      components: [P6Tabs],
      html: `<p6-tabs></p6-tabs>`,
    });

    expect(page.root).toEqualHtml(`
      <p6-tabs>
        <mock:shadow-root>
         </mock:shadow-root>
      </p6-tabs>
    `);
  });

  it('Should display the third tab', async () => {
    const page = await newSpecPage({
      components: [P6Tabs],
      html: `
      <p6-tabs selected="third">
        <div id="first" title="What is Lorem Ipsum?">1</div>
        <div id="second" title="Where does it come from?">2</div>
        <div id="third" title="Why do we use it?">3</div>
        <div id="fourth" title="Where can I get some?">4</div>
      </p6-tabs>
      `,
    });

    expect(page.root).toMatchSnapshot();
  });
});
