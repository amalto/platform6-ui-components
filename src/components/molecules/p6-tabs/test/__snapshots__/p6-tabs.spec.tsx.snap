// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-tabs Should display the third tab 1`] = `
<p6-tabs selected="third">
  <mock:shadow-root>
    <div class="tabs">
      <ul>
        <li class="has-tooltip-arrow has-tooltip-bottom" data-tooltip="What is Lorem Ipsum?">
          <a href="#first" id="first-tab">
            <span class="title">
              What is Lorem Ipsum?
            </span>
            <span aria-hidden="true" class="delete"></span>
          </a>
        </li>
        <li class="has-tooltip-arrow has-tooltip-bottom" data-tooltip="Where does it come from?">
          <a href="#second" id="second-tab">
            <span class="title">
              Where does it come from?
            </span>
            <span aria-hidden="true" class="delete"></span>
          </a>
        </li>
        <li class="has-tooltip-arrow has-tooltip-bottom" data-tooltip="Why do we use it?">
          <a class="is-active" href="#third" id="third-tab">
            <span class="title">
              Why do we use it?
            </span>
            <span aria-hidden="true" class="delete"></span>
          </a>
        </li>
        <li class="has-tooltip-arrow has-tooltip-bottom" data-tooltip="Where can I get some?">
          <a href="#fourth" id="fourth-tab">
            <span class="title">
              Where can I get some?
            </span>
            <span aria-hidden="true" class="delete"></span>
          </a>
        </li>
      </ul>
    </div>
    <div class="tab-content">
      3
    </div>
  </mock:shadow-root>
  <div id="first" title="What is Lorem Ipsum?">
    1
  </div>
  <div id="second" title="Where does it come from?">
    2
  </div>
  <div id="third" title="Why do we use it?">
    3
  </div>
  <div id="fourth" title="Where can I get some?">
    4
  </div>
</p6-tabs>
`;
